s64da_benchmark_toolkit/reporting.py

```diff
@@ -0,0 +1,172 @@
+
+import logging
+import os
+
+from csv import writer as csv_writer
+from datetime import datetime
+from pathlib import Path
+from uuid import uuid4
+
+import pandas
+
+from natsort import index_natsorted, order_by_index
+from tabulate import tabulate
+
+from .correctness import Correctness, CorrectnessResult
+from .netdata import Netdata
+
+
+LOG = logging.getLogger()
+
+
+class QueryMetric:
+    dataframe_columns = (
+        'stream_id', 'query_id', 'timestamp_start', 'timestamp_stop',
+        'runtime', 'status')
+
+    def __init__(self, *, stream_id, query_id, timestamp_start, timestamp_stop,
+                 status, result, plan):
+        self.stream_id = stream_id
+        self.query_id = query_id
+        self.timestamp_start = datetime.fromtimestamp(timestamp_start)
+        self.timestamp_stop = datetime.fromtimestamp(timestamp_stop)
+        self.status = status
+        self.result = result
+        self.plan = plan
+
+    def make_file_name(self, extension):
+        return f'{self.stream_id}_{self.query_id}.{extension}'
+
+    @property
+    def dataframe(self):
+        runtime = (self.timestamp_stop - self.timestamp_start).total_seconds()
+        return pandas.DataFrame(data=[[
+            self.stream_id, self.query_id, self.timestamp_start,
+            self.timestamp_stop, runtime, self.status
+        ],], columns=QueryMetric.dataframe_columns)
+
+
+class Reporting:
+    def __init__(self, benchmark, args, config):
+        self.uuid = uuid4()
+        self.benchmark = benchmark
+        self.results_root_dir = 'results'
+        self.html_output = os.path.join(self.results_root_dir, 'report.html')
+        self.query_results = os.path.join(self.results_root_dir, 'query_results')
+        self.explain_analyze_dir = os.path.join(self.results_root_dir, 'query_plans')
+        self.scale_factor = args.scale_factor
+        self.explain_analyze = args.explain_analyze
+        self.all_query_metrics = []
+        self.netdata_output_file = args.netdata_output_file
+        self.config = config
+
+        self.output = args.output
+        self.csv_file = args.csv_file
+        if 'csv' in self.output:
+            Path(self.csv_file).touch()
+
+        if args.check_correctness and not os.path.exists(self.results_root_dir):
+            try:
+                os.makedirs(self.results_root_dir)
+            except OSError as exc:
+                LOG.exception(f'Could not create directory {self.results_root_dir}')
+
+    def run_report(self, reporting_queue):
+        df = pandas.DataFrame(columns=QueryMetric.dataframe_columns)
```

Comment:
Why not make df a `self` variable so you don't have to pass it all over the place?

Reply:
Because of recommendations on how to work with DataFrames in pandas, i.e. they should not necessarily be modified in place.

Reply:
Wouldn't it make more sense then to have the df  `self.df` and when you need to modify it, modify it into a new local variable? Like in `_sort_df`, for example.

Reply:
Hm, I am not really convinced. I see the benefit of having `self.df` but I tend to rather pass it around following the pandas semantics. `_sort_df` does that as well.

Reply:
I find it make the intention of the code not very clear. Because, after all, even passing the data frame as a function parameter doesn't make a new copy of it, and if you modify that df in-place (which is an option for some of panda's functions) then the original will change as well. At least that's how I understand it.
Making it into `self.df` will make that intention clear, and when you want to run a function of the df that doesn't change the original, you'll assign it to a new local variable. Just like you do in `_sort_df`: `df = df.reindex(index=order_by_index(df.index, index_sort))`

Reply:
I honestly don't like making it a `self.df` and still prefer passing it around. But I will make the change.

Reply:
Maybe to put more rationale in here: I like the idea of immutability. I know that pandas itself does not always adhere to it, but most of the time. However, since the code follows one path only, I can live with the change to `self.df`.